### ISSUE TITLE
Make CI report failure fast

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#11099]
When submitting a PR or pushing an update to it the CI by Travis and AppVeyor is triggered automatically. In particular AppVeyor can take long time. 
Each of the two tools always executes the full list of builds configured for it, and only after completing the whole list the success/failure icon at the bottom of the PR discussion GitHub page is updated.

I consider this a waste of time and computational resources.
Why not let (at least by default) Travis and AppVeyor terminate as soon as one of its builds fails?

I'm assuming here that usually there is only one mistake that makes one or more builds fail,
and if there are more mistakes it usually helps to fix one of them and then retry the CI builds.

**Comment by levitte** [comment]
Do you know how to do this? I don't, I thought that the jobs were independent of each other...

**Comment by beldmit** [comment]
Is it a good idea?
We can save a lot of time in (rare?) cases when the openssl cannot be built at all and more often case when doc nits are found. Otherwise, it can add more time waist than time economy...

**Comment by FdaSilvaYY** [comment]
I made a try, a long time ago, but someone raise the objection that it will "add more complexity".
I may open a PR, if there is some interest now ?

**Comment by DDvO** [comment]
> Do you know how to do this? I don't, 

Unfortunately I don't either - I have virtually no experience with CI technology.

> I thought that the jobs were independent of each other...

They must be because they are started (at least partially) in parallel,
but I assume for each CI invocation there is an overall controller process that has overview over the builds and can kill/abort all builds as soon as one of them fails.

**Comment by FdaSilvaYY** [comment]
See #2692

**Comment by DDvO** [comment]
> We can save a lot of time in (rare?) cases when the openssl cannot be built at all and more often case when doc nits are found. Otherwise, it can add more time waist than time economy...

I don't see were a time waste occurs, except in the very rare case that there are multiple independent mistakes and the user checks all failed builds and fixes all of the errors at once.

**Comment by DDvO** [comment]
> See #2692.
> Travis has probably improve their API since.

This looks like it would be rather easy to do :)

**Comment by FdaSilvaYY** [comment]
One of the CI caveats (I regret), is that you cannot stop the Builds of your own PR (on OpenSSL) when you see a few of them are failing. unless you are a project member ;)

**Comment by DDvO** [comment]
> One of the CI caveats (I regret), is that you cannot stop the Builds of your own PR (on OpenSSL) when you see that a few of them are failing. unless you are a project member ;)

What you _could_ do - as far as you realize this situation at all (which unfortunately, as written above, is not flagged on the PR discussion GitHub page as long as the list of builds is not finished)  and in case you can quickly come up with a fix - is to push a new version, which restarts the CI.

**Comment by mattcaswell** [comment]
On 15/02/2020 15:04, David von Oheimb wrote:
>     We can save a lot of time in (rare?) cases when the openssl cannot
>     be built at all and more often case when doc nits are found.
>     Otherwise, it can add more time waist than time economy...
> 
> I don't see were a time waste occurs, except in the very rare case that
> there are multiple independent mistakes and the user checks all failed
> builds and fixes all of the errors at once.

I regularly check all the build fails and find more than one mistake in
a single run.

Matt

**Comment by mspncp** [comment]
Breaking early might be useful if CI failures on the master branch are rare, but currently that's often not the case. If the master branch is not stable, an early CI break strategy might be counterproductive:

The reason is because CI failures on master also affect the pull requests, because they are merged with master before being built. Currently, we often check all CI failures and if they are unrelated we allow the pull request to be merged nevertheless. If the CIs abort on the first error, we loose this ability.

**Comment by slontis** [comment]
You would have to keep going if a build timed out (represented by the ! icon) as opposed to an actual error since these happen quite often..

**Comment by DDvO** [comment]
> I regularly check all the build fails and find more than one mistake in a single run.

My typical experience when submitting PRs is that there is a minor issue, e.g., with `util/libcrypto.num` or `find-doc-nits` that makes most of the builds fail. As soon as one build fails for this reason it makes little sense to keep the CI running.

At least I usually notice a failed CI run only after Travis has finished, usually after many failed build attempts, and I find it too cumbersome to manually check the logs of all failed builds to try and find a potential further mistake after spotting one of the usual culprits.

**Comment by DDvO** [comment]
> Breaking early might be useful if CI failures on the master branch are rare, but currently that's often not the case. If the master branch is not stable, an early CI break strategy might be counterproductive:
> 
> The reason is because CI failures on master also affect the pull requests, because they are merged with master before being built. Currently, we often check all CI failures and if they are unrelated we allow the pull request to be merged nevertheless. If the CIs abort on the first error, we loose this ability.

> You would have to keep going if a build timed out (represented by the ! icon) as opposed to an actual error since these happen quite often..

Sigh, yes, currently Travis often times out at least for one build for no real reason.

So let me refine my suggestion:

Make (by default) CI runs fail as soon as one build fails except due to timeout.

**Comment by levitte** [comment]
Find-doc-nits isn't run in all jobs, only those with the environment variable CHECKDOCS defined. So stopping *everything* just because the documentation checker made *some* jobs fail is not a good idea at all. Some of us do bother to look at all failing jobs, since they cover different build aspects.

**Comment by DDvO** [comment]
Meanwhile I also have experienced myself the value of getting reported potentially different errors from more than one CI build.
So I've further adapted the title of this PR to "Make CI report failure fast", asking for a smaller change:

Change the color of the respective status indicator from yellow to red as soon as one of the Travis CI or one of the AppVeyor builds fails.
This way the owner of the PR can notice quickly that something has gone wrong, without having to click and view the Details pages.

**Comment by nhorman** [comment]
Given that we no longer use travis and appveyor, I'm assuming this is no longer relevant

Marking as inactive to be closed at the end of 3.4 dev, barring further input

**Comment by t8m** [comment]
I am closing this outright. This is outdated and IMO irrelevant.
